lib/source_file.cpp

```diff
@@ -69,7 +71,35 @@ void InputSources::addComment(SourceInfo srcInfo, bool singleLine, cstring body)
     if (!singleLine)
         // Drop the "*/"
         body = body.exceptLast(2);
-    comments.push_back(new Comment(srcInfo, singleLine, body));
+    auto srcPos = SourcePosition(srcInfo.getStart());
+
+    comments.push_back(new Comment(srcPos, singleLine, body));
+}
+
+std::string SourceInfo::getComments() const {
+    if ((sources == nullptr)) {
+        return "";
+    }
+
+    std::ostringstream oss;
+    uint targetLine = start.getLineNumber();
+    const Comment *closestComment = nullptr;
+
+    for (const auto &comment : sources->comments) {
+        uint commentLine = comment->getSourcePosition().getLineNumber();
+        auto diff = (targetLine - commentLine);
+        if (commentLine < targetLine &&
+            (closestComment == nullptr ||
+             commentLine > closestComment->getSourcePosition().getLineNumber()) &&
+            (diff == 0 || diff == 1)) {
+            closestComment = comment;
+        }
+    }
+
+    if (closestComment != nullptr) {
+        oss << closestComment->toString();
```

Comment:
What is the point of this terrible code? You're using `std::ostringstream` just to do conversion from `cstring` to `std::string`? This at least involves multiple string copies: first into the stream, and then back. Consider rewriting the code in more straightforward manner, e.g.:

```c++
return closestComment ? closestComment->toString().string() : "";
```